src/election.ts

```diff
@@ -0,0 +1,282 @@
+      .if(this.leaseId, 'Create', '==', this._leaderRevision)
+      .then(this.namespace.put(this.leaseId).value(value).lease(this.leaseId))
+      .commit();
+
+    if (!r.succeeded) {
+      this._leaderKey = '';
+      return this.emitOrThrowError(new EtcdNotLeaderError());
+    }
+  }
+
+  public async resign() {
+    if (!this.isCampaigning) {
+      return;
+    }
+
+    const r = await this.namespace
+      .if(this.leaseId, 'Create', '==', this._leaderRevision)
+      .then(this.namespace.delete().key(this.leaseId))
+      .commit();
+
+    if (!r.succeeded) {
+      // If fail, revoke lease for performing resigning
+      await this.lease.revoke();
+      this.lease = this.namespace.lease(this.ttl);
+      this.lease.on('lost', () => this.onLeaseLost());
+      this.leaseId = '';
+    }
+
+    this._leaderKey = '';
+    this._leaderRevision = '';
+    this._isCampaigning = false;
+  }
+
+  public async getLeader() {
+    const result = await this.namespace.getAll().sort('Create', 'Ascend').keys();
+    if (result.length === 0) {
+      throw new EtcdNoLeaderError();
+    }
+    return `${this.getPrefix()}${result[0]}`;
+  }
+
+  public getPrefix() {
+    return `${Election.prefix}/${this.name}/`;
+  }
+
+  private async waitForElected() {
+    // find last create before this
+    const lastRevision = new BigNumber(this.leaderRevision).minus(1).toString();
+    const result = await this.namespace
+      .getAll()
+      .maxCreateRevision(lastRevision)
+      .sort('Create', 'Descend')
+      .keys();
+
+    // no one before this, elected
+    if (result.length === 0) {
+      return;
+    }
+
+    // wait all keys created ealier are deleted
+    await waitForDeletes(this.namespace, result);
+  }
+
+  private async observe() {
+    if (this._isObserving) {
+      return;
+    }
+
+    try {
+      this._isObserving = true;
+
+      // looking for current leader
+      let leaderKey = '';
+      const result = await this.namespace.getAll().sort('Create', 'Ascend').keys();
+
+      if (result.length === 0) {
+        // if not found, wait for leader
+        const watcher = await this.parent.watch().prefix(this.getPrefix()).create();
+        try {
+          leaderKey = await new Promise<string>((resolve, reject) => {
+            watcher.on('put', kv => resolve(kv.key.toString()));
+            watcher.on('error', reject);
+          });
+        } finally {
+          await watcher.cancel();
+        }
+      } else {
+        leaderKey = `${this.getPrefix()}${result[0]}`;
+      }
+
+      // emit current leader
+      this.emit('leader', leaderKey);
+
+      // wait for delete event
+      await waitForDelete(this.parent, leaderKey);
+    } finally {
+      this._isObserving = false;
+    }
+
+    // only keep watch if listened
+    if (this.listenerCount('leader') > 0) {
+      this.tryObserve();
+    }
+  }
+
+  private tryObserve(): void {
+    this.observe().catch(error => this.emitOrThrowError(error));
```

Comment:
Wouldn't this always lead to uncaught rejections if no handler this attached?, all `tryObserve` calls should be awaited.

Reply:
Is it proper to `observe` again (next tick or a little delay) after caught an error always? I worry about any situation I didn't consider.